code/hash/Hacl.Hash.SHA3.fsti

```diff
@@ -9,22 +9,61 @@ module Hacl.Hash.SHA3
 // implementation in code/sha3 does not export the "update_multi" version, since
 // it does everything in one go, so it's convenient to make this explicit here.)
 //
-// Final remark: this module is completely noextract, so it generates no code at run-time.
+// NOTE: unlike other modules, this one is not entirely noextract
+// inline_for_extraction. There are two reasons. First, because all Keccak
+// variants share the same state type, these functions do *NOT* need to be
+// inlined to fit in the Low* subset. Second, for this reason, they are not
+// always reduced at compile-time for a chosen value of `a`, meaning that we
+// need the code to look decent (and not have everything inlined aggressively
+// when there are no opportunities for reduction).
 
 open Spec.Hash.Definitions
 open Hacl.Hash.Definitions
 
 inline_for_extraction noextract
-val init (a: sha3_alg { not (is_shake a) }): init_st (|a, ()|)
+val init (a: sha3_alg): init_st (|a, ()|)
 
 inline_for_extraction noextract
-val update_multi (a: sha3_alg { not (is_shake a) }): update_multi_st (|a, ()|)
+val update_multi (a: sha3_alg): update_multi_st (|a, ()|)
 
 inline_for_extraction noextract
-val update_last (a: sha3_alg { not (is_shake a) }): update_last_st (|a, ()|)
+val update_last (a: sha3_alg): update_last_st (|a, ()|)
 
 inline_for_extraction noextract
 val finish (a: sha3_alg { not (is_shake a) }): finish_st (| a, ()|)
 
 inline_for_extraction noextract
 val hash (a: sha3_alg { not (is_shake a) }): hash_st a
+
+/// A couple helpers specifically for the Keccak functor, which live here
+/// because this module has an fsti and therefore can friend specs.
+
+let v_len a (l: Lib.IntTypes.size_t): Spec.Hash.Definitions.output_length a =
+  let _ = allow_inversion hash_alg in
+  if is_shake a then
+    Lib.IntTypes.(v #U32 #PUB (l <: size_t))
+  else
+    ()
+
+module B = LowStar.Buffer
+module ST = FStar.HyperStack.ST
+
+/// This is a variant of Hacl.Hash.SHA3.finish that takes an optional output
+/// length. This is used in the Keccak streaming functor, wherein this signature
+/// eventually becomes exposed in the C API. As a result, we cannot have `l` be
+/// an indexed type (e.g. if is_shake then size_t else unit) because that would
+/// not be extractable to C. So, we contend with a suboptimal contract, which
+/// is: "if a is a shake algorithm, then the length is ignored".
```

Comment:
Isn't it the opposite, i.e. if a is *not* a shake algorithm, then the length is ignored?
Also, why do you need the length precondition as a refinement instead of in the requires clause?

Reply:
Good catch thanks